k2/csrc/array_of_ragged.h

```diff
@@ -0,0 +1,191 @@
+     when the NumSrcs() is small.  [Also: see GetRowInfoMulti()].
+   */
+  // TODO: implement it...
+  Array2<int32_t> MetaRowSplits();
+
+  // could POSSIBLY add this so this code could be used in functions like
+  // Stack(). would be like MetaRowSplits but with an extra 1st row containing
+  // 0,1,2,... We could perhaps create it with 1 extra initial row so this is
+  // always convenient to output.
+  // TODO: implement it...
+  Array2<int32_t> Offsets();
+
+  /*
+    Returns the meta-row-splits for a particular axis, with 0 <= axis <
+    NumAxes(); this is the cumulative sum of the TotSize(axis) for all of the
+    sources, with MetaRowSplits(axis).Dim() == NumSrcs() + 1.
+
+    Note: in ragged_ops.cu we refer to this as composed_row_splits
+  */
+  // TODO: implement it...
+  Array1<int32_t> MetaRowSplits(int32_t axis);
+
+  /* Return the device-accessible meta-row-ids, which are the row-ids
+     corresponding to MetaRowSplits(); this tells us, for indexes into the
+     appended/concatenated array, which source array they belong to, i.e.
+     elements are in [0,NumSrcs()-1].
+
+     This cannot be an Array2 because unlike the MetaRowSplits(), all the
+     row-ids arrays are of different lengths.
+
+     Note: in ragged_ops.cu we refer to this as composed_row_ids.
+  */
+  // TODO: implement it...
+  Array1<int32_t *> MetaRowIds();
+
+  /*
+    Returns the meta-row-ids for a particular axis, with 0 <= axis < NumAxes();
+    this is the row-ids corresponding to MetaRowSplits(axis), and its elements
+    gives, for indexes into the concatentated shape (concatenated on axis 0),m
+    which source they come from.  E.g. element 100 of MetaRowIds(2)
+    would tell us which source an idx012 with value 100 into axis 2 of
+    concatenated array would come from.
+  */
+  // TODO: implement it...
+  Array1<int32_t> MetaRowIds(int32_t axis);
+
+ private:
+  ContextPtr c_;
+  int32_t num_srcs_;
+  int32_t num_axes_;
+  Array2<int32_t *> row_splits_;  // shape [num_axes_ - 1][num_srcs_]
+  Array2<int32_t *> row_ids_;     // shape [num_axes_ - 1][num_srcs_]
+  Array1<int32_t> tot_sizes_;     // dim num_axes_, this is on CPU
+};
+
+/*
+  ArrayOfRagged<T> is a 1-dimensional array of Ragged<T>.
+  It is intended for situations where you want to do some operations on
+  arrays of ragged arrays, without explicitly concatenating them (e.g. to
+  save time).   This is a fairly low-level interface, intended to
+  be used mostly by CUDA/C++ implementation code.  It is a convenience
+  wrapper that saves you the trouble of creating arrays of pointers.
+ */
+template <typename T>
+struct Array1OfRagged {
+  Array1OfRaggedShape shape;
+
+  // Array of the individual values pointers of the source arrays, indexed by
+  // shape
+  Array1<T *> values;
+
+  int32_t NumSrcs() { return values.Dim(); }
+  ContextPtr &Context() { return shape.Context(); }
+
+  Array1OfRagged() = default;
+
+  Array1OfRagged(Ragged<T> *srcs, int32_t num_srcs) {
```

Comment:
Would be good if there is some document saying: All entries in `srcs` should have the same number of axes.